core/store/src/trie/mem/resharding.rs

```diff
@@ -0,0 +1,347 @@
+    root_id: MemTrieNodeId,
+    children: &ChildrenView<'a, M>,
+    mut value: Option<&ValueView>,
+    key_nibbles: Vec<u8>,
+    intervals_nibbles: &[Range<Vec<u8>>],
+    updates_tracker: &mut UpdatesTracker,
+) -> Option<OldOrUpdatedNodeId> {
+    let mut new_children = [None; 16];
+    let mut changed = false;
+
+    if !intervals_nibbles.iter().any(|interval| interval.contains(&key_nibbles)) {
+        value = None;
+        changed = true;
+    }
+
+    for i in 0..16 {
+        let Some(child) = children.get(i) else {
+            continue;
+        };
+
+        let child_key_nibbles = [key_nibbles.clone(), vec![i as u8]].concat();
+        let new_child = retain_multi_range_recursive(
+            child,
+            child_key_nibbles,
+            intervals_nibbles,
+            updates_tracker,
+        );
+        match new_child {
+            Some(OldOrUpdatedNodeId::Old(id)) => {
+                new_children[i] = Some(OldOrUpdatedNodeId::Old(id));
+            }
+            Some(OldOrUpdatedNodeId::Updated(id)) => {
+                changed = true;
+                new_children[i] = Some(OldOrUpdatedNodeId::Updated(id));
+            }
+            None => {
+                changed = true;
+                new_children[i] = None;
+            }
+        }
+    }
+
+    if changed {
+        let new_node = UpdatedMemTrieNode::Branch {
+            children: Box::new(new_children),
+            value: value.map(|v| v.to_flat_value()),
+        };
+        // TODO(#12074): squash the branch if needed.
+        // TODO(#12074): return None if needed.
+        Some(OldOrUpdatedNodeId::Updated(updates_tracker.add_node(new_node)))
+    } else {
+        Some(OldOrUpdatedNodeId::Old(root_id))
+    }
+}
+
+/// Based on the key and the intervals, makes decision on the subtree exploration.
+fn retain_decision(key: &[u8], intervals: &[Range<Vec<u8>>]) -> RetainDecision {
+    let mut should_descend = false;
+    for interval in intervals {
+        if key < interval.start.as_slice() {
+            if interval.start.starts_with(key) {
+                should_descend = true;
+            } else {
+                // No retain for this interval.
+            }
+        } else {
+            if key >= interval.end.as_slice() {
+                // No retain for this interval.
+            } else if interval.end.starts_with(key) {
+                should_descend = true;
+            } else {
+                return RetainDecision::RetainAll;
+            }
+        }
+    }
+
+    if should_descend {
+        RetainDecision::Descend
+    } else {
+        RetainDecision::NoRetain
+    }
+}
+
+// TODO(#12074): tests for
+// - multiple retain ranges
+// - removing keys one-by-one gives the same result as corresponding range retain
+// - `retain_split_shard` API
+// - all results of squashing branch
+// - checking not accessing not-inlined nodes
+// - proof correctness
+// - (maybe) retain result is empty or complete tree
```

Comment:
Sorry, what does this test mean? Also could we add a test where we are checking no nodes are getting deleted (decreasing refcount) from the trie?

Reply:
It means the same what you said :) Changed wording a bit